From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SupeChicken666 <me@supechicken666.dev>
Date: Wed, 11 Feb 2026 22:32:27 +0800
Subject: [PATCH 17/33] init: Allow overriding read-only props on waydroid-init

Signed-off-by: SupeChicken666 <me@supechicken666.dev>
---
 init/property_service.cpp | 13 +++++++++++--
 1 file changed, 11 insertions(+), 2 deletions(-)

diff --git a/init/property_service.cpp b/init/property_service.cpp
index d3d422e..42cc46d 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -391,6 +391,8 @@ class PersistWriteThread {
 
 static std::optional<uint32_t> PropertySet(const std::string& name, const std::string& value,
                                            SocketConnection* socket, std::string* error) {
+    static bool waydroid_init_done = false;
+    static pid_t waydroid_init_pid = -1;
     size_t valuelen = value.size();
 
     if (!IsLegalPropertyName(name)) {
@@ -412,12 +414,19 @@ static std::optional<uint32_t> PropertySet(const std::string& name, const std::s
         if (pi != nullptr) {
             // ro.* properties are actually "write-once", unless the system decides to
             if (StartsWith(name, "ro.") && !weaken_prop_override_security &&
-                !StartsWith(name, "ro.boot.vbmeta.")) {
+                !StartsWith(name, "ro.boot.vbmeta.") &&
+                (waydroid_init_done || socket ? socket->cred().pid != waydroid_init_pid : true)) {
                 *error = "Read-only property was already set";
                 return {PROP_ERROR_READ_ONLY_PROPERTY};
             }
 
             __system_property_update(pi, value.c_str(), valuelen);
+        } else if (!waydroid_init_done && socket && name == "waydroid.init.start") {
+            waydroid_init_pid = socket->cred().pid;
+            return {PROP_SUCCESS};
+        } else if (!waydroid_init_done && socket && name == "waydroid.init.done") {
+            waydroid_init_done = true;
+            return {PROP_SUCCESS};
         } else {
             int rc = __system_property_add(name.c_str(), name.size(), value.c_str(), valuelen);
             if (rc < 0) {
@@ -1410,7 +1419,7 @@ void PropertyLoadBootDefaults() {
     load_properties_from_partition("odm", /* support_legacy_path_until */ 28);
     load_properties_from_partition("product", /* support_legacy_path_until */ 30);
     load_properties_from_file("/vendor/waydroid.prop", nullptr, &properties);
- 
+
     if (int vendor; ParseInt(properties["ro.vendor.build.version.sdk"], &vendor) &&
         vendor >= 33) {
         if (properties["ro.hardware.egl"] == "swiftshader") {
